src/AndersonAcceleration.f90

```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
+      DEALLOCATE(solver%Gx)
+      DEALLOCATE(solver%r)
+      DEALLOCATE(solver%tmpvec)
+    ENDIF
+    DEALLOCATE(solver%alpha)
+    CALL solver%LS%clear()
+    solver%s=0
+    solver%n=-1
+    solver%depth=-1
+    solver%start=0
+    solver%depth=-1
+    solver%beta=0.0_SRK
+    solver%isInit=.FALSE.
   ENDIF
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  CALL Anderson_Destroy(solver%id)
-#endif
-  solver%isInit=.FALSE.
+
 ENDSUBROUTINE clear_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Single step of the Anderson acceleration
-!> @param solver The anderson solver to act on
-!>
-!> This routine takes a single anderson acceleration step
+!> @brief Performs a single step of Anderson Acceleration acting on the input solution vector.
+!>        If depth is set to zero, or if the iteration count is below the Anderson starting
+!>        point the behavior is to under-relax the solution using the mixing parameter (beta)
+!>        as the under-relaxation factor.
+!> @param solver Anderson solver to take step with
+!> @param x_new New solution iterate and under-relaxed / accelerated return vector
 !>
-SUBROUTINE step_AndersonAccelerationType(solver)
+SUBROUTINE step_AndersonAccelerationType(solver,x_new)
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  IF(solver%iter==0) THEN
-    CALL Anderson_Reset(solver%id)
+  CLASS(VectorType),INTENT(INOUT) :: x_new
+
+  INTEGER(SIK) :: i,k,depth_s
+  REAL(SRK) :: tmpA,tmpb
+
+  REQUIRE(x_new%n == solver%n)
+
+  !Update iteration counter
+  solver%s=solver%s+1
+
+  IF(solver%s >= solver%start) THEN
+    depth_s=MIN(solver%depth,solver%s-solver%start)
+    !Push back vectors to make room for new ones
+    DO i=depth_s,1,-1
+      CALL BLAS_copy(solver%Gx(i),solver%Gx(i+1))
+      CALL BLAS_copy(solver%r(i),solver%r(i+1))
+    ENDDO
+
+    !Set new vectors based on input
+    CALL BLAS_copy(x_new,solver%Gx(1))
+    CALL BLAS_copy(solver%Gx(1),solver%r(1))
+    CALL BLAS_axpy(solver%x(1),solver%r(1),-1.0_SRK)
+
+    !Get fit coefficients
+    IF(depth_s == 1) THEN
+      !Depth 1 is an especially simple case, where alpha can be calculated with a simple formula
+      CALL BLAS_copy(solver%r(2),solver%tmpvec)
+      CALL BLAS_axpy(solver%r(1),solver%tmpvec,-1.0_SRK)
+      tmpA=BLAS_dot(solver%tmpvec,solver%tmpvec)
+      tmpb=BLAS_dot(solver%r(2),solver%tmpvec)
+      solver%alpha(1)=tmpb/tmpA
+    ELSEIF(depth_s > 1) THEN
+      !Construct coefficient matrix, right hand side, and solve for fit coefficients
+      DO i=1,depth_s
+        CALL BLAS_copy(solver%r(i),x_new)
+        CALL BLAS_axpy(solver%r(depth_s+1),x_new,-1.0_SRK)
+        tmpb=BLAS_dot(solver%r(depth_s+1),x_new)
+        CALL solver%LS%b%set(i,tmpb)
+        DO k=1,i
+          CALL BLAS_copy(solver%r(depth_s+1),solver%tmpvec)
+          CALL BLAS_axpy(solver%r(k),solver%tmpvec,-1.0_SRK)
+          tmpA=BLAS_dot(solver%tmpvec,x_new)
```

Comment:
bugs me that all the BLAS stuff are subroutines except this one... but I know we can't do anything about that